messages/src/msg_fields/protocols/cred_issuance/v1/mod.rs

```diff
@@ -0,0 +1,225 @@
+            }
+            CredentialIssuanceTypeV1_0::CredentialPreview => {
+                Err(utils::not_standalone_msg::<D>(kind_str))
+            }
+        }
+    }
+
+    fn delayed_serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::OfferCredential(v) => MsgWithType::from(v).serialize(serializer),
+            Self::ProposeCredential(v) => MsgWithType::from(v).serialize(serializer),
+            Self::RequestCredential(v) => MsgWithType::from(v).serialize(serializer),
+            Self::IssueCredential(v) => MsgWithType::from(v).serialize(serializer),
+            Self::Ack(v) => MsgWithType::from(v).serialize(serializer),
+            Self::ProblemReport(v) => MsgWithType::from(v).serialize(serializer),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct CredentialPreviewV1 {
+    #[serde(rename = "@type")]
+    msg_type: CredentialPreviewV1MsgType,
+    pub attributes: Vec<CredentialAttr>,
+}
+
+impl CredentialPreviewV1 {
+    pub fn new(attributes: Vec<CredentialAttr>) -> Self {
+        Self {
+            msg_type: CredentialPreviewV1MsgType,
+            attributes,
+        }
+    }
+}
+
+/// Non-standalone message type.
+/// This is only encountered as part of an existent message.
+/// It is not a message on it's own.
+#[derive(Copy, Clone, Debug, Default, Deserialize, PartialEq)]
+#[serde(try_from = "CowStr")]
+struct CredentialPreviewV1MsgType;
+
+impl<'a> From<&'a CredentialPreviewV1MsgType> for CredentialIssuanceTypeV1_0 {
+    fn from(_value: &'a CredentialPreviewV1MsgType) -> Self {
+        CredentialIssuanceTypeV1_0::CredentialPreview
+    }
+}
+
+impl<'a> TryFrom<CowStr<'a>> for CredentialPreviewV1MsgType {
+    type Error = String;
+
+    fn try_from(value: CowStr) -> Result<Self, Self::Error> {
+        let value = MessageType::try_from(value.0.as_ref())?;
+
+        if let Protocol::CredentialIssuanceType(CredentialIssuanceKind::V1(
+            CredentialIssuanceTypeV1::V1_0(_),
+        )) = value.protocol
+        {
+            if let Ok(CredentialIssuanceTypeV1_0::CredentialPreview) =
+                CredentialIssuanceTypeV1_0::from_str(value.kind)
+            {
+                return Ok(CredentialPreviewV1MsgType);
+            }
+        }
+
+        Err(format!("message kind is not {}", value.kind))
+    }
+}
+
+impl Serialize for CredentialPreviewV1MsgType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let protocol = Protocol::from(CredentialIssuanceTypeV1_0::parent());
+        let kind = CredentialIssuanceTypeV1_0::from(self);
+        format_args!("{protocol}/{}", kind.as_ref()).serialize(serializer)
+    }
+}
+
+transit_to_aries_msg!(
+    OfferCredentialV1Content: OfferCredentialV1Decorators,
+    CredentialIssuanceV1
+);
```

Comment:
This macro has variadic arguments. You can append `CredentialIssuance` at the end to make the conversions chain longer:

```suggestion
transit_to_aries_msg!(
    OfferCredentialV1Content: OfferCredentialV1Decorators,
    CredentialIssuanceV1, CredentialIssuance
);
```

I'll remove the macro at some point when I find the time to work more on the `transitive` crate so it accepts generics. It will become redundant then.

Reply:
ahhhhh thank you, this makes sense. nice macro!